batch2/batch/driver/scheduler.py

```diff
@@ -39,57 +81,75 @@ def __init__(self, app):
+    ''', (user,))
+            async for record in records:
+                should_wait = False
+                await unschedule_job(self.app, record)
 
         return should_wait
 
     async def schedule_1(self):
-        records = self.db.execute_and_fetchall(
-            '''
-SELECT job_id, batch_id, spec, cores_mcpu,
-  ((jobs.cancelled OR batches.cancelled) AND NOT always_run) AS cancel,
-  userdata, user
-FROM jobs
-INNER JOIN batches ON batches.id = jobs.batch_id
-WHERE jobs.state = 'Ready' AND batches.closed
-LIMIT 50;
-''')
+        fair_share = await self.compute_fair_share()
+
+        users = fair_share.keys()
+        random.shuffle(users)
 
         should_wait = True
-        async for record in records:
-            batch_id = record['batch_id']
-            job_id = record['job_id']
-            id = (batch_id, job_id)
-
-            if record['cancel']:
-                log.info(f'cancelling job {id}')
-                await mark_job_complete(self.app, batch_id, job_id, None,
-                                        'Cancelled', None, None, None, 'cancelled')
-                should_wait = False
-                continue
-
-            i = self.inst_pool.healthy_instances_by_free_cores.bisect_key_left(record['cores_mcpu'])
-            if i < len(self.inst_pool.healthy_instances_by_free_cores):
-                instance = self.inst_pool.healthy_instances_by_free_cores[i]
-                assert record['cores_mcpu'] <= instance.free_cores_mcpu
-                log.info(f'scheduling job {id} on {instance}')
-                try:
-                    await schedule_job(self.app, record, instance)
-                except Exception:
-                    log.exception(f'while scheduling job {id} on {instance}')
-                should_wait = False
+
+        for user in users:
+            scheduled_cores_mcpu = 0
+            allocated_cores_mcpu = fair_share[user]
+
+            records = self.db.execute_and_fetchall(
+                '''
+    SELECT job_id, batch_id, spec, cores_mcpu,
+      ((jobs.cancelled OR batches.cancelled) AND NOT always_run) AS cancel,
+      userdata, user
+    FROM jobs
+    INNER JOIN batches ON batches.id = jobs.batch_id
+    WHERE jobs.state = 'Ready' AND batches.closed AND batches.user = %s
+    LIMIT 50;
+    ''', (user, ))
+
+            async for record in records:
+                batch_id = record['batch_id']
+                job_id = record['job_id']
+                id = (batch_id, job_id)
+
+                if record['cancel']:
+                    log.info(f'cancelling job {id}')
+                    await mark_job_complete(self.app, batch_id, job_id, None,
+                                            'Cancelled', None, None, None, 'cancelled')
+                    should_wait = False
+                    continue
+
+                if scheduled_cores_mcpu + record['cores_mcpu'] > allocated_cores_mcpu:
+                    log.info(f'skipping job {id}; exceeded fair share limits')
```

Comment:
I'd either make this a break, or take out the log.  The scheduler runs often and this will generate a lot of messages (when the cluster is full, basically 50/user).